pdf/model/optimize/image_ppi.go

```diff
@@ -0,0 +1,202 @@
+/*
+ * This file is subject to the terms and conditions defined in
+ * file 'LICENSE.md', which is part of this source code package.
+ */
+
+package optimize
+
+import (
+	"fmt"
+	"image"
+	"math"
+
+	"github.com/unidoc/unidoc/pdf/contentstream"
+	. "github.com/unidoc/unidoc/pdf/core"
+	"github.com/unidoc/unidoc/pdf/model"
+	"golang.org/x/image/draw"
+)
+
+// ImagePPI optimizes images.
+// It implements interface model.Optimizer.
+// Does not support inline images yet.
+type ImagePPI struct {
+	ImageUpperPPI float64
+}
+
+func scaleImage(stream *PdfObjectStream, scale float64) error {
```

Comment:
When handling the error up from scaleImage, would be better to output a Debug ERROR message and leave object unmodified.